src/utils/threading.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+
+namespace util {
+
+/**
+ * @brief Implementation of a countdown latch for threading synchronization.
+ *
+ * This has roughly the same interface as std::latch, but that requires C++20,
+ * which we do not support.
+ *
+ * @see https://en.cppreference.com/w/cpp/thread/latch
+ */
+class latch {
+public:
+	/**
+	 * @brief Create a new latch.
+	 *
+	 * @param count This is the number of countdowns before the latch is unlocked.
+	 */
+	latch(std::size_t count);
+
+	latch(const latch&) = delete;
+
+	latch& operator=(const latch&) = delete;
+
+	/**
+	 * @brief Wait until the latch is unlocked.
+	 */
+	void wait() const;
+
+	/**
+	 * @brief Wait until the latch is unlocked, with a timeout.
+	 *
+	 * @param dur The timeout duration.
+	 * @return bool True if the latch was unlocked, false if timed out.
+	 */
+	template <typename Rep, typename Period>
+	bool wait_for(const std::chrono::duration<Rep, Period>& dur) const {
+		std::unique_lock lock(mutex);
+		return cv.wait_for(lock, dur, [&]() { return count <= 0; });
+	}
+
+	/**
+	 * @brief Wait until the latch is unlocked until a specific timepoint.
+	 *
+	 * @param tp The timepoint to wait until.
+	 * @return bool True iff the latch was unlocked before the given time.
```

Comment:
Yeah, this is the same as `wait_for()` except:
* It's often convenient to wait until some time instead of for a duration
* It waits using the clock associated with the timepoint instead of the default